pandas/tools/merge.py

```diff
@@ -273,9 +273,17 @@ def _get_join_info(self):
                                                  sort=self.sort, how=self.how)
 
             if self.right_index:
-                join_index = self.left.index.take(left_indexer)
+                if len(self.left) > 0:
+                    join_index = self.left.index.take(left_indexer)
+                else:
+                    join_index = self.right.index.take(right_indexer)
+                    left_indexer = np.array([np.nan] * len(join_index))
             elif self.left_index:
-                join_index = self.right.index.take(right_indexer)
+                if len(self.right) > 0:
+                    join_index = self.right.index.take(right_indexer)
+                else:
+                    join_index = self.left.index.take(left_indexer)
+                    right_indexer = np.array([np.nan] * len(join_index))
```

Comment:
It is used for filling when ``left_on`` is specified. When use ``np.array([-1])*len(join(index))`` or leave it as it is, the result will be:

```
import pandas as pd
left = pd.DataFrame([], columns=['a', 'b', 'c'])
right = pd.DataFrame([[1, 2, 3], [4, 5, 6], [7, 8, 9]],
                             columns=['x', 'y', 'z'])
result = pd.merge(left, right, how='right', left_on='a', right_index=True)
result
#    a    b    c  x  y  z
# 0  0  NaN  NaN  1  2  3
# 1  1  NaN  NaN  4  5  6
# 2  2  NaN  NaN  7  8  9
```